### PR TITLE
[c#] Support for options

### PR DESCRIPTION
**Pull request by yowl**
This PR adds support to c# codegen for `option<>`

Redoes #819 

A large change for just enabling options, but this also implies adding support for anonymous types.  I also hit problems with namespace clashes, so reworked some of the namespace logic to upper camel case more names.

I've largely copied Java for the lower/lifting, but we have a class for each option type.  We cannot use `Nullable<T>` in C# because in wit `option<option...` is possible, but `Nullable` is only available over value types so `Nullable<Nullable` is not valid.  A generic `Option<T>` should be possible as a code size optimisation.

Split the options codegen test so that simple options can be brought up without support for `result`.

cc @alexcrichton  for the the split of the option test

### CODE REVIEW

**Comment by alexcrichton** [comment]
Splitting the test up looks good to me! I'll leave it to the others for the C# review too

File: crates/csharp/src/lib.rs [hunk 1]
```diff
@@ -207,7 +224,7 @@ impl WorldGenerator for CSharp {
         _files: &mut Files,
     ) -> Result<()> {
         let name = &format!("{}-world", resolve.worlds[world].name);
-        let mut gen = self.interface(resolve, name, false);
+        let mut gen = self.interface(resolve, name, false, false);
```
Comment (jsturtevant) [inline crates/csharp/src/lib.rs:227]: could we alias the true/false to something?  It is hard to know what the intent is here

**Review by jsturtevant — COMMENTED** [review]
LGTM, Thanks for improving the namespacing!

File: crates/csharp/src/lib.rs (hunk 1 above)
Comment (yowl) [reply crates/csharp/src/lib.rs]: good call, in fact the second one was no longer used, and we can use the existing `Direction` for the first one.  Have made that change

Comment (yowl) [reply crates/csharp/src/lib.rs]: edit, ok, that flag is needed, have made them both enums to make the code a bit more readable.

**Comment by yowl** [comment]
Hi, if this is ok, can @jsturtevant  or @silesmo approve it?

**Review by jsturtevant — APPROVED** [review]

**Comment by yowl** [comment]
Thanks!

**Pull request merged**